C++ API parity: Dropout, Dropout2d, Dropout3d

pbelevich opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#29761 C++ API parity: Dropout, Dropout2d, Dropout3d**

Differential Revision: [D18530820](https://our.internmc.facebook.com/intern/diff/D18530820)

pbelevich committed:
C++ API parity: Dropout, Dropout2d, Dropout3d

[ghstack-poisoned]

yf225 commented on torch/csrc/api/include/torch/nn/options/dropout.h:
```diff
@@ -7,14 +7,41 @@
+
+using Dropout3dFuncOptions = DropoutFuncOptions;
+
+}
```
nit:
```suggestion
} // namespace functional
```

yf225 commented on torch/csrc/api/include/torch/nn/options/dropout.h:
```diff
@@ -7,14 +7,41 @@
+  /// can optionally do the operation in-place. Default: False
+  TORCH_ARG(bool, inplace) = false;
+
+  TORCH_ARG(bool, training) = false;
```
Based on https://github.com/pytorch/pytorch/blob/3fb9bbc99b8e1af286ca84045697a951f5346832/torch/nn/functional.py#L788, it seems that the default value for `training` should be `true`

yf225 replied:
also it would be awesome to move `TORCH_ARG(bool, training) = false;` above `TORCH_ARG(bool, inplace) = false;`, to be consistent with the argument order of the Python functional `dropout` :D

yf225 commented on torch/csrc/api/include/torch/nn/modules/dropout.h:
```diff
@@ -14,66 +14,90 @@ namespace torch {
+    TORCH_CHECK(
+        options.p() >= 0. && options.p() <= 1.,
+        "dropout probability has to be between 0 and 1, but got ",
+        options.p());
```
We can probably move the check to the `reset()` function and call `reset()` from the constructor, which would be more consistent with the design of the other C++ modules :D

yf225 commented on torch/csrc/api/include/torch/nn/functional/dropout.h:
```diff
@@ -0,0 +1,81 @@
+
+} // namespace detail
+
+inline Tensor dropout(Tensor &input, 
```
nit:
```suggestion
inline Tensor dropout(Tensor& input, 
```

yf225 commented on torch/csrc/api/include/torch/nn/functional/dropout.h:
```diff
@@ -0,0 +1,81 @@
+
+} // namespace detail
+
+inline Tensor dropout2d(Tensor &input, 
```
nit:
```suggestion
inline Tensor dropout2d(Tensor& input, 
```

yf225 commented on torch/csrc/api/include/torch/nn/functional/dropout.h:
```diff
@@ -0,0 +1,81 @@
+
+} // namespace detail
+
+inline Tensor dropout3d(Tensor &input, 
```
nit:
```suggestion
inline Tensor dropout3d(Tensor& input, 
```

yf225 commented on torch/csrc/api/include/torch/nn/functional/dropout.h:
```diff
@@ -0,0 +1,81 @@
+
+} // namespace functional
+} // namespace nn
+} // namespace torch
```
nit: we might need a new line here

yf225 commented on torch/csrc/api/include/torch/nn/modules/dropout.h:
```diff
@@ -14,66 +14,90 @@ namespace torch {
-/// See the documentation for `FeatureDropoutImpl` class to learn what methods
-/// it provides, or the documentation for `ModuleHolder` to learn about
-/// PyTorch's module storage semantics.
-TORCH_MODULE(FeatureDropout);
```
I think we might want to deprecate `FeatureDropout` for one release and then remove it in the next one. For 1.4 (which will be branch cut on 11/20) we can deprecate it first by printing a warning message in the `FeatureDropout` constructor asking people to move to `Dropout2d` / `Dropout3d`, and in the 1.5 release we can remove `FeatureDropout` completely :D

yf225 reviewed (commented):
@pbelevich Thanks a lot for the awesome work! I left some comments

pbelevich committed:
Update on "C++ API parity: Dropout, Dropout2d, Dropout3d"




[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: Dropout, Dropout2d, Dropout3d"




[ghstack-poisoned]

yf225 commented on torch/csrc/api/src/nn/modules/dropout.cpp:
```diff
@@ -8,42 +9,71 @@
+  : detail::_DropoutNd<FeatureDropoutImpl>(options_) {
+  TORCH_WARN("torch::nn::FeatureDropout module is deprecated."
+             "Use Dropout{2,3}d instead.");
+}
```
It would be awesome to test for the deprecation message, similar to https://github.com/pytorch/pytorch/blob/eef349a679f5331e05de3c9462e285d1ab99c8a8/test/cpp/api/modules.cpp#L1176-L1188

yf225 commented on torch/csrc/api/include/torch/nn/modules/dropout.h:
```diff
@@ -14,34 +14,95 @@ namespace torch {
+/// See the documentation for `Dropout3dImpl` class to learn what methods it
+/// provides, or the documentation for `ModuleHolder` to learn about PyTorch's
+/// module storage semantics.
+TORCH_MODULE(Dropout3d);
```
I believe we can update the parity status of `Dropout` / `Dropout2d` / `Dropout3d` in `test/cpp_api_parity/parity-tracker.md` as well :D

pbelevich committed:
Update on "C++ API parity: Dropout, Dropout2d, Dropout3d"




[ghstack-poisoned]

yf225 reviewed (approved):
@pbelevich Thanks so much for the awesome work!

pbelevich committed:
Update on "C++ API parity: Dropout, Dropout2d, Dropout3d"




[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: Dropout, Dropout2d, Dropout3d"


Differential Revision: [D18530820](https://our.internmc.facebook.com/intern/diff/D18530820)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: Dropout, Dropout2d, Dropout3d"


Differential Revision: [D18530820](https://our.internmc.facebook.com/intern/diff/D18530820)

[ghstack-poisoned]
